translators/src/Print/Idris.hs

```diff
@@ -1,110 +1,155 @@
+{-# Language OverloadedStrings #-}
 module Print.Idris
   ( printModule
+  , render
   , runIdris
   ) where
 
-import Data.List (intercalate)
+import Prettyprinter
+import Prettyprinter.Render.String (renderString)
 
 import Grammar
-import Print.Generic
 
-import_, univ, arr, typedel, assign :: String
-import_ = "import "
-univ = "Type"
-arr = " -> "
-typedel = " : "
-assign = " = "
+newtype Idris ann = Idris {get :: Doc ann}
 
-printImport :: Import -> String
-printImport (ImportLib VecMod) = import_ ++ "Data.Vect"
+-- to be migrated
+class Keywords rep where
+  import_ :: rep
+  assign  :: rep
+  data_   :: rep
+  rec     :: rep
+  univ    :: rep
+  arr     :: rep
+
+instance Keywords (Doc ann) where
+  import_ = "import"
+  assign  = "="
+  data_   = "data"
+  rec     = "record"
+  univ    = "Type"
+  arr     = "->"
+
+class TypeAnn rep where
+  typeAnn :: rep -> rep -> rep
+  teleCell :: rep -> rep -> rep
+
+instance TypeAnn (Doc ann) where
+  typeAnn trm typ = trm <+> ":" <+> typ
+  teleCell trm typ = parens $ trm <+> ":" <+> typ
+
+printImport :: Import -> Doc ann
+printImport (ImportLib VecMod) = import_ <+> "Data.Vect"
 -- There rest are builtin
-printImport (ImportLib NatMod) = ""
-printImport (ImportLib StringMod) = ""
-printImport (ImportLib ListMod) = ""
+printImport (ImportLib NatMod) = mempty
```

Comment:
This is the cause of the empty lines between the `module` header and the rest of the module. The `Monoid` instance for `Doc` uses [emptyDoc](https://hackage.haskell.org/package/prettyprinter-1.7.1/docs/Prettyprinter.html#v:emptyDoc), which has a height of 1.

If we want to avoid these, we should change the signature of `printImport` to be something like
```haskell
printWithImport :: Import -> Doc ann -> Doc ann
```

and then just return the `Doc ann` unchanged if we don't need to add an import.